ENH: Add PGI flang compiler support for Windows

xoviat opened the pull request:
Closes #9940.

xoviat commented:
One issue with this is that flang supports OSX and linux as well (it's LLVM, after all), but the linker and such is different.

xoviat commented:
Sorry for the noise here. I'm actually not using the CI here, so I'll close for now.

xoviat commented on numpy/distutils/fcompiler/pg.py:
```diff
@@ -54,10 +56,57 @@ def get_flags_linker_so(self):
+        }
+
+    compile_switch = '-nomp'
+    object_switch = '-o '     # No space after /Fo!
```
Once thing I'm not clear about is that the object switch and the compile switch are the same for flang. I just put a fake compile switch here (pretty harmless option).

isuruf replied:
Isn't it `-c` ?

xoviat replied:
See https://ci.appveyor.com/project/xoviat/scipy/build/1.0.23#L3924

isuruf replied:
`-c` is given twice here, `"C:\Miniconda36-x64\Library\bin\flang.exe -Ic:\python36-x64\lib\site-packages\numpy\core\include -c -c  scipy\fftpack\src\dfftpack\dcosqb.f -obuild\temp.win-amd64-3.6\scipy\fftpack\src\dfftpack\dcosqb.o"`

xoviat commented:
@charris Questions? Assuming tests pass, okay to merge.

xoviat commented:
Just some additional changes to improve autoconfiguration for conda.

pv commented:
Why does it inherit from PGroupFCompiler when it's a different compiler 
and you redefine all attributes? Maybe just put it in a new file.

xoviat commented:
I think it's fine in the same file, but agree with the class hierarchy. Will change it.

xoviat committed:
ENH: distutils: improve windows autoconfiguration

Autoconfigure libraries and headers when conda is detected
on the path. Also support copenblas and clapack when found.

xoviat committed:
ENH: disutils: add the flang compiler

Flang is the first CRT-compatbile compiler for windows.
It's added for Python versions with a compatible ABI.
Future work may include adding it for other platforms.

xoviat commented:
@charris Commits have been cleaned up.

charris commented:
Needs rebase.

xoviat commented:
@charris Done.

dimpase commented:
it looks as this this branch is a bit too Windows-centric.  It particular checks for Python version looks strange.

It is perfectly possible to build numpy with flang (and clang) on FreeBSD 11 (where flang is ported, i.e. you just to `pkg install flang` to install it), and (as far as I know, I didn't test it myself) on Linux with flang provided by conda. (this is with Python 2.7.14).

xoviat commented:
> it looks as this this branch is a bit too Windows-centric. It particular checks for Python version looks strange.

Uh yes. It would tend to be windows-centric since *nix already has gfortran but windows doesn't (at least not without hacks, which the entire point of this is to avoid).

> and (as far as I know, I didn't test it myself) on Linux with flang provided by conda. (this is with Python 2.7.14).

Uh yes. I'm aware of that since I'm a maintainer of the conda package. But why? gfortran already works with the native ABI and has better performance at this time.

> It particular checks for Python version looks strange.

flang will *never* support Python 2.7 on windows. So why put in the effort to write outdated Python code?

I mean, I have nothing against using flang on *nix. But the principle benefit is for windows. Because windows has required major hacks for far too long.

dimpase commented:
On Mon, Nov 13, 2017 at 2:06 AM, xoviat <notifications@github.com> wrote:

> it looks as this this branch is a bit too Windows-centric. It particular
> checks for Python version looks strange.
>
> Uh yes. It would tend to be windows-centric since *nix already has
> gfortran but windows doesn't (at least not without hacks, which the entire
> point of this is to avoid).
>
If you work with an LLVM toolchain then you might not want to mix it with
GCC toolchain, for one or another reason,
on Windows or not; e.g. on OSX it is always iffy whether each new release
of Xcode will let you build gfortran, for instance Xcode 9 cannot build
gfortran 6 or 5, it can only build gfortran 7.

and (as far as I know, I didn't test it myself) on Linux with flang
> provided by conda. (this is with Python 2.7.14).
>
> Uh yes. I'm aware of that since I'm a maintainer of the conda package. But
> why? gfortran already works with the native ABI and has better performance
> at this time.
>
Can you debug gfortran-compiled code with lldb? Or you need  gdb?

> It particular checks for Python version looks strange.
>
> flang will *never* support Python 2.7 on windows. So why put in the
> effort to write outdated Python code?
>
I'm merely worried that this branch might break things that already work in
Python 2.7.
(Besides, are we really talking about some tricky, unportable by simple
`from ___future___ include ...` statements Python3 code that one has to
write here?)

> I mean, I have nothing against using flang on *nix. But the principle
> benefit is for windows. Because windows has required major hacks for far
> too long.
>
speaking of *nices, OSX is not easy to deal with.

Last but not the least, while still just a conjecture, we might see CUDA
features available in flang at some point, as it has NVIDIA people working
on it.

xoviat commented:
> If you work with an LLVM toolchain then you might not want to mix it with
GCC toolchain, for one or another reason,
on Windows or not; e.g. on OSX it is always iffy whether each new release
of Xcode will let you build gfortran, for instance Xcode 9 cannot build
gfortran 6 or 5, it can only build gfortran 7.

IMHO flang should not be used for *nix because the performance is currently not up to par.

> Can you debug gfortran-compiled code with lldb? Or you need  gdb?

You need gdb.

> I'm merely worried that this branch might break things that already work in
Python 2.7.
(Besides, are we really talking about some tricky, unportable by simple
`from ___future___ include ...` statements Python3 code that one has to
write here?)

This is not going to break anything that already works with Python 2.7. What this is going to do is open up new options for windows users on Python 3. There is a very good reason that this is being done this way.

> speaking of *nices, OSX is not easy to deal with.

I don't understand this. Compiling scientific projects is just a `brew install gcc` away. I don't use OSX except for on Travis-CI, as I'm a windows and linux user, but surely running commands locally cannot be more difficult?

What I think is this: you want the flang compiler to work on *nix with Python 2.7. But let me point out two points:

1. I'm not going to do this work because I don't see any benefit. I'm not going to prevent anyone else from submitting their own PR, but I don't see a need at least from my POV.
2. It is really not a good idea to be using Python 2.7. I mean really. Nick (the CPython developer) explains this better than I could hope to. Numpy is dropping support. But people will continue to use it as it is almost an ideological commitment at this point, and then they'll be surprised when they can't get new features.

dimpase commented:
On Tue, Nov 14, 2017 at 12:43 AM, xoviat <notifications@github.com> wrote:

> If you work with an LLVM toolchain then you might not want to mix it with
> GCC toolchain, for one or another reason,
> on Windows or not; e.g. on OSX it is always iffy whether each new release
> of Xcode will let you build gfortran, for instance Xcode 9 cannot build
> gfortran 6 or 5, it can only build gfortran 7.
>
> IMHO flang should not be used for *nix because the performance is
> currently not up to par.
>
Often one just wants something that works, even if it's 3 times slower...


> Can you debug gfortran-compiled code with lldb? Or you need gdb?
>
> You need gdb.
>
See? Here it starts. E.g. you have clang-built Python with gfortan-built
extension, and
it really puts you in a non-man land as soon as something is really not fine
between the two.


> I'm merely worried that this branch might break things that already work in
> Python 2.7.
> (Besides, are we really talking about some tricky, unportable by simple
> from ___future___ include ... statements Python3 code that one has to
> write here?)
>
> This is not going to break anything that already works with Python 2.7.
> What this is going to do is open up new options for windows users on Python
> 3. There is a very good reason that this is being done this way.
>
> speaking of *nices, OSX is not easy to deal with.
>
> I don't understand this. Compiling scientific projects is just a brew
> install gcc away. I don't use OSX except for on Travis-CI, as I'm a
> windows and linux user, but surely running commands locally cannot be more
> difficult?
>

FYI, Homebrew seriously messes up the OSX native toolchain. It's either
Homebrew all the way,
or not at all. That's why e.g. sagemath does not use it. (It's not as
drastic as mingw in Windows world,
but not too far, if you ask me).

What I think is this: you want the flang compiler to work on *nix with
> Python 2.7. But let me point out two points:
>
>    1. I'm not going to do this work because I don't see any benefit. I'm
>    not going to prevent anyone else from submitting their own PR, but I don't
>    see a need at least from my POV.
>    2. It is really not a good idea to be using Python 2.7. I mean really.
>    Nick (the CPython developer) explains this better than I could hope to.
>    Numpy is dropping support. But people will continue to use it as it is
>    almost an ideological commitment at this point, and then they'll be
>    surprised when they can't get new features.
>
> I am all for moving to Python 3, it's just that it is a lot of work for
big Python 2 projects, and often you really need something that works for
Python 2 and 3 to be able to figure out the right way to change the code to
work in Python 3.
Surely we at sagemath need to be in sync with numpy and scipy, but we are
still quite a way from working Python 3
support.

xoviat commented:
I want to be clear: I'm not against adding support for other platforms. I simply don't have time to add support myself. In addition, it will not be able to efficiently test the code at this time. So if you want, go ahead after this PR is merged; it shouldn't be that much work. I just won't be able to do it myself, and I don't believe that PRs should be blocked unless they cause regressions (or have the potential to cause regressions in complex parts of the codebase without substantial review). This PR is tested to work with windows, and even if it's broken (which it isn't), no one will be the worse.

@charris Can I get a 1.14 milestone on this?

insertinterestingnamehere commented:
FWIW I agree that Linux/OSX support with flang would be nice to have in general, but it shouldn't be a blocker here. Windows is where we need the additional compiler the most since existing c runtime incompatibilities make using gfortran tricky.

xoviat commented:
@charris @rgommers AFAIK this is ready.

xoviat commented:
Rebased on master.

xoviat commented:
@charris Ping.

xoviat commented:
@charris Can this please go into 1.14?

xoviat commented:
@charris Can we please get this in?

charris commented:
Sure, I don't see how it can hurt. If problems turn up when it starts getting used we'll just need to fix them. Thanks @xoviat .

xoviat commented:
Yeah, IMO it shouldn't be **too** risky, even at this point, because there's no existing functionality changed and there's no C code.
